language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,362 @@
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum Code {
+    Native,
+    Declared(CodeUnit),
+}
+
+pub struct CodeUnit {
+    // ... max_stack_size is gone...
+
+    // locals carry only the locals signature without the arguments.
+    // However there is a single instruction that access locals uniformly,
+    // that is, as if there was a single vector of arguments + locals.
+    // The compiler (and generally clients) are supposed to do the math properly
+    pub locals: SignatureIndex,
+    pub code: Vec<Bytecode>,
+}
+
+//
+// Bytecodes
+//
+
+// The main issue to digest here is that every bytecode that deals with generic is now
+// split in 2 versions. The non generic and the generic. That has some advantages
+// we have talked about and we may touch on more
+pub enum Bytecode {
+    Pop,
+    Ret,
+    Nop, // should we have one?
+
+    // the branch instructions pending a structured control flow model we may want to provide
+    BrTrue(CodeOffset),
+    BrFalse(CodeOffset),
+    Branch(CodeOffset),
+
+    // load value on stack
+    LdU8(u8),
+    LdU64(u64),
+    LdU128(u128),
+    LdU8Vector(U8VectorPoolIndex),
+    LdAddr(AddressPoolIndex),
+    LdTrue,
+    LdFalse,
+
+    // random operators
+    CastU8,
+    CastU64,
+    CastU128,
+    Add,
+    Sub,
+    Mul,
+    Mod,
+    Div,
+    Shl,
+    Shr,
+    BitOr,
+    BitAnd,
+    Xor,
+    Or,
+    And,
+    Not,
+    Lt,
+    Gt,
+    Le,
+    Ge,
+
+    // system ops
+    Abort,
+    GetTxnSenderAddress,
+
+    // locals ops
+    CopyLoc(LocalIndex),
+    MoveLoc(LocalIndex),
+    StLoc(LocalIndex),
+    MutBorrowLoc(LocalIndex),
+    ImmBorrowLoc(LocalIndex),
+
+    // reference ops
+    ReadRef,
+    WriteRef,
+    FreezeRef,
+
+    // equality ops, reference and not
+    Eq,
+    Neq,
+
+    // calls
+    Call(FunctionHandleIndex),
+    CallGeneric(FunctionSpecIndex),
+
+    // fields
+    MutBorrowField(FieldHandleIndex),
+    MutBorrowFieldGeneric(FieldSpecIndex),
+    ImmBorrowField(FieldHandleIndex),
+    ImmBorrowFieldGeneric(FieldSpecIndex),
+
+    // type ops
+    Pack(StructDefinitionIndex),
+    PackGeneric(StructSpecIndex),
+    Unpack(StructDefinitionIndex),
+    UnpackGeneric(StructSpecIndex),
+    MutBorrowGlobal(StructDefinitionIndex),
+    MutBorrowGlobalGeneric(StructSpecIndex),
+    ImmBorrowGlobal(StructDefinitionIndex),
+    ImmBorrowGlobalGeneric(StructSpecIndex),
+    Exists(StructDefinitionIndex),
+    ExistsGeneric(StructSpecIndex),
+    MoveFrom(StructDefinitionIndex),
+    MoveFromGeneric(StructSpecIndex),
+    MoveToSender(StructDefinitionIndex),
+    MoveToSenderGeneric(StructSpecIndex),
+}
+
+//
+// Constants (possible story)
+//
+
+// A ConstantPool will replace all data pools (address, byte_array)
+// with a more general constant pool that would allow a mix of primitive types and
+// Vectors of them. Types can and should be restricted to allow for only a subset
+// of types.
+pub struct ConstantPool {
+    type_: Type,
+    data: Vec<u8>,
```

Comment:
this will not be the runtime representation which will instead be a `Vec<Value>`.
However that means nothing at the binary format level so I am keeping it here as `u8`.
That brings back this whole ordeal of using this as the runtime representation in the entire codebase which in one form or another we should probably break.
In that respect I even wonder if we should have a file (an `.rs` file) which good fidelity to the binary format and Rust abstraction that will have mapping to that.... 
don't know...
thoughts?

Reply:
But as a whole I like the idea. I think it gives us a good bit of leeway and future proofing for not a lot of effort

Reply:
let me try to expand on what I meant and definitely curious about your take (and others as well).

Overall I keep seeing this tension between what a binary format is and what a Rust representation for that format can be.
We mixed the 2 in the past and I think that has proved to be a problematic story. This `Vec<u8>` vs `Value` for constants is a poster child of the problem, but the whole story around the header (common vs data segment and what not) is a similar problem. The binary format is a set of compressed tables and that is it. In that respect I wonder if a Rust representation that expresses that would be nice to have. And then whatever higher level representation is needed we will define for those needs and translate that to the binary format.
I suppose I am questioning the entire use of `CompiledModule` and `CompiledScript` here.

In any case I'll try to see if I can come up with something a bit more concrete soon

Reply:
>  And then whatever higher level representation is needed we will define for those needs and translate that to the binary format.

This seems like a good story. The file format would be as high fidelity as possible (as compared to the actual serialized/binary format). The high level models are the point where shit should be sorted out into nice Rust code. 
I think the reason it was so mixed is because the system of views+file format was/is a mess

> I suppose I am questioning the entire use of CompiledModule and CompiledScript here.

What does this mean exactly?